rcl/src/rcl/timer.c

```diff
@@ -0,0 +1,239 @@
+    return now_ret;  // rcl error state should already be set.
+  }
+  uint64_t previous_ns =
+    rcl_atomic_exchange_uint64_t(&timer->impl->last_call_time, now_steady.nanoseconds);
+  uint64_t since_last_call = now_steady.nanoseconds - previous_ns;
+  rcl_timer_callback_t typed_callback =
+    (rcl_timer_callback_t)rcl_atomic_load_uintptr_t(&timer->impl->callback);
+  typed_callback(timer, since_last_call);
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_timer_is_ready(const rcl_timer_t * timer, bool * is_ready)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(timer, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(is_ready, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(timer->impl, "timer is invalid", return RCL_RET_TIMER_INVALID);
+  int64_t time_until_next_call;
+  rcl_ret_t ret = rcl_timer_get_time_until_next_call(timer, &time_until_next_call);
+  if (ret != RCL_RET_OK) {
+    return ret;  // rcl error state should already be set.
+  }
+  *is_ready = (time_until_next_call <= 0) && !rcl_atomic_load_bool(&timer->impl->canceled);
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_timer_get_time_until_next_call(const rcl_timer_t * timer, int64_t * time_until_next_call)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(timer, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(time_until_next_call, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(timer->impl, "timer is invalid", return RCL_RET_TIMER_INVALID);
+  rcl_steady_time_point_t now;
+  rcl_ret_t ret = rcl_steady_time_point_now(&now);
+  if (ret != RCL_RET_OK) {
+    return ret;  // rcl error state should already be set.
+  }
+  uint64_t period = rcl_atomic_load_uint64_t(&timer->impl->period);
+  *time_until_next_call =
+    (rcl_atomic_load_uint64_t(&timer->impl->last_call_time) + period) - now.nanoseconds;
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_timer_get_time_since_last_call(const rcl_timer_t * timer, uint64_t * time_since_last_call)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(timer, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(time_since_last_call, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(timer->impl, "timer is invalid", return RCL_RET_TIMER_INVALID);
+  rcl_steady_time_point_t now;
+  rcl_ret_t ret = rcl_steady_time_point_now(&now);
+  if (ret != RCL_RET_OK) {
+    return ret;  // rcl error state should already be set.
+  }
+  *time_since_last_call =
+    (now.nanoseconds - rcl_atomic_load_uint64_t(&timer->impl->last_call_time));
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_timer_get_period(const rcl_timer_t * timer, uint64_t * period)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(timer, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(period, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_FOR_NULL_WITH_MSG(timer->impl, "timer is invalid", return RCL_RET_TIMER_INVALID);
+  *period = rcl_atomic_load_uint64_t(&timer->impl->period);
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_timer_exchange_period(const rcl_timer_t * timer, uint64_t new_period, uint64_t * old_period)
```

Comment:
But to be fair, I don't mention that this function is atomic (in the docstring), and so it seems that your familiarity with atomics may have lead you to assume this function was atomic, not the other way around. The word exchange in the name means exactly that, you give me a new value and I'll give the old, but doesn't mean to imply atomic behavior, though I understand what you mean that it invokes that idea based on the name.